### DISCUSSION TITLE
New activation functions

### DISCUSSION THREAD

**Discussion by mariodeflorio** [leakec/tfc#22] [Q&A] [answered]
Hey Carl!
I want to try to use different activation functions for the neural networks, to see if we can improve the performance for certain PDEs with discontinuity. I tried to implement them in the source codes, but I had some issues. Is it straightforward to implement new activation functions? If so, can you provide a short tutorial on how to do it? 
Thanks!

**Comment by leakec** [comment, marked as answer]
Hi Mario!
Sure thing. 

## Python-only implementation
It is likely easiest to prototype your functions in `src/tfc/utils/BF/BF_Py.py`; this is the Python implementation of the basis functions. For ELM-based activation functions, you can create a univariate version that inherits from the `ELM` class. Your new univariate activation function should implement the `_Hint` method, which is used to calculate the basis function matrix and its derivatives. See `ELMTanh`  in that file for an example. If you instead want a multivariate version, then create a class that inherits from `nELM` and implements the `_nHint` method. Similar to the univariate case, `_nHint` is used to calculate the basis function matrix and its derivatives. See `nELMTanh` in that file for an example. 

In Python, we sometimes use JAX directly to calculate the derivatives of our basis functions. This is the case in `ELMTanh` and `nELMTanh`. Here, a recursive function is used to calculate the correct derivative based on the input `d`.

The Python-only implementations do not have a C++ interface, so they can't be wrapped up and packaged into the JIT XLA compiler. Hence, they cannot be used in the `jit` transformation. However, TFC comes with `pejit`, which aggressively caches values based on their inputs, e.g., you can cache `x` and all functions that depend on `x` as compile time constants with `pejit`. `pejit` is really easy to use, and very similar to JIT. You can find a tutorial on how to use the Python-only basis functions [here](https://tfc-documentation.readthedocs.io/en/latest/notebooks/BasisFunctionBackends.html), and a tutorial on `pejit` [here](https://tfc-documentation.readthedocs.io/en/latest/notebooks/pejit.html)

Next, you'll need to add your functions to the `utfc` and/or `mtfc` basis class. In the `__init__` methods of these classes there will be a block of code that sets the `self.basisClass` attribute based on the `self.basis` value. Add the logic here to use your basis function backend when appropriate.

Once you implement your functions, you'll need to install them from the development version you're working on. There is a folder called `utils` in the top-level directory that has a `Makefile`. You can call `make install` to install TFC from your local git repo. Under the hood, this essentially does:
1. Uninstalling the version of TFC you already have.
2. Running `python setup.py bdist_wheel` to create a binary wheel for your current machine.
3. Running `pip install` on the binary wheel that was created in 2.

## C++ implementation
Once you have something you are happy with, it would awesome if we could implement it completely and merge it with main so others can use it too. To do this, we will need a C++ implementation as well. This is done with two files: `src/tfc/utils/BF/BF.h` and `src/tfc/utils/BF/BF.cxx`. 

To implement the univariate version, create a class that inherits from `ELM` and implements the `Hint` method: see `ELMTanh` for an example. For the multivariate version, create a class that inherits from `nELM` and implements the `nElmHint` method: see `nELMTanh` for an example. 

Like the Python version, you'll need to install to make this available (and add the basis function option to `utfc` and `mtfc`, though you likely already did that above if you implemented the Python-only version first). The installation steps are the same as in the Python version. When you run these steps, the installation automatically uses `swig` to create a Python wrapper for the C++ versions so these are available through the Python interface.
